### ISSUE TITLE
Array OOB access should be a compile error

### ISSUE THREAD

**Issue by mohammadfawaz** [FuelLabs/sway#2923]
E.g. 
```rust
script;

fn main() -> u32 {
    let ary = [1, 2, 3, 4];
    let idx = 45;
    ary[idx]
}
```
This compiles fine but it shouldn't, at least in situations where `idx` is constant evaluatable.

**Comment by otrho** [comment]
This surprises me considering https://github.com/FuelLabs/sway/tree/master/test/src/e2e_vm_tests/test_programs/should_fail/array_oob
The only difference is the indirection via the `idx` var.  Obviously it's enough.

**Comment by mohammadfawaz** [comment]
Yeah `idx` is considered a dynamic index, unless we run the const-evaluator on it.

**Comment by esdrubal** [comment]
I checked the code that returns the OOB access compiler error [sway-core/src/ir_generation/function.rs#L1688](https://github.com/FuelLabs/sway/blob/13ffbc3328805b8bf1abed1d34f9ecc597efa045/sway-core/src/ir_generation/function.rs#L1688).

In the case of the example above, the index_expr is compiled to an Instruction::Load with the value Instruction::GetPointer().
Maybe we can check the current block for an Instruction::Store of a const to the same pointer.

**Comment by mohammadfawaz** [comment]
@esdrubal one way to do this is to leverage the const evaluator in https://github.com/FuelLabs/sway/blob/master/sway-core/src/ir_generation/const_eval.rs.
* If the const evaluator passes on the index, then we can just replace the value with the evaluated constant and do the OOB checks on it.
* If the const evaluator fails (it can only do so much, and not every expression is const-evaluatable), then we just don't do anything.

**Comment by esdrubal** [comment]
@mohammadfawaz Thanks for pointing me out to [const_eval.rs](https://github.com/FuelLabs/sway/blob/master/sway-core/src/ir_generation/const_eval.rs). While using compile_constant_expression_to_constant I noticed that it is able to compile to a constant when index is a Literal. When it is a VariableExpression it doesn't know about variables in the scope and may try to compile to the value of a global const over shadowing local variables.

I am working on a solution for this and will send a PR later today.
